gcc: support alternate mechanism for providing GCC rpaths

sethrj opened the pull request:
Spack's GCC install hardwires (through the specs file) the linker to adding an rpath entry to GCC's library dir, so that compiled binaries get the right libstdc++ etc. Unfortunately this leaves build tools like CMake unable to automatically remove those hardcoded paths when creating a relocatable installation.

My solution is to slightly refactor the way spack modifies the GCC `specs` file so that a simple robust spec file:
```
*link_libgcc_rpath:


```
and the linker option `--specs=norpath.spec` will skip the addition of the spack rpaths. (You can instead use CMake's `CMAKE_BUILD_RPATH` to provide the paths to the GCC libraries as needed). I've tested that the updated spec file works exactly like the original with single-line builds of an executable, as well as using GCC to generate shared libraries, and that the spec option above can disable the extra rpaths in both cases.

Fixes #26582 .

This *also* fixes the rpath construction on macOS, which has always been incorrect: rpaths must be given as separate rpath commands to the linker, rather than combined with a colon as is the case for linux.

sethrj committed:
gcc: support runtime ability to not install spack rpaths

Fixes #26582 .

sethrj committed:
gcc: Fix malformed specs file and add docs

The updated docs point out that the spack-modified GCC does *not*
follow the usual behavior of LD_RUN_PATH!

sethrj commented:
I've tested this out and it works on both GCC4 and GCC11. Furthermore, I realized that since the inception of the gcc spack package, spack-compiled GCC has been silently changing the behavior of the GCC-driven linker: **Spack GCC ignores the `LD_RUN_PATH` environment variable because `-rpath` is always being passed to the linker flags!**

I don't know what kind of unintended consequence this has had. Maybe some packages have required manual rpath flags instead? Maybe it's why we define module prefixes for `LD_LIBRARY_PATH` rather than `LD_RUN_PATH`, because the latter gets ignored?

@tgamblin @alalazo Maybe this is a known behavior to you, but it should probably be documented (it is now), and it might be worth making a `~rpath` variant that defines a runtime `LD_RUN_PATH` append rather than a hidden linker flag, so that downstream/user codes that expect LD_RUN_PATH to work will get that behavior.

With the default spack-built GCC compiler:
```console
$ echo 'int main() { return 0; }' > derp.cc
$ LD_RUN_PATH=/my/ld/runpath g++ derp.cc -o derp.exe && readelf -d derp.exe | grep rpath
 0x000000000000000f (RPATH)              Library rpath: [/projects/spack/opt/spack/gcc-4.8.5/gcc/bny6lal/lib:/projects/spack2020/opt/spack/gcc-4.8.5/gcc/bny6lal/lib64]
```
after the modifications in the PR, passing the user-defined `norpath.spec` defined above:
```console
$ LD_RUN_PATH=/my/ld/runpath g++ --specs=norpath.spec derp.cc -o derp.exe && readelf -d derp.exe | grep rpath
 0x000000000000000f (RPATH)              Library rpath: [/my/ld/runpath]
```

michaelkuhn commented:
Have you checked how this interacts with applications actually using one of the libraries on GCC's lib directory (such as `libasan` or `libgomp`)? Will they still be found?

sethrj commented:
> Have you checked how this interacts with applications actually using one of the libraries on GCC's lib directory (such as `libasan` or `libgomp`)? Will they still be found?

You're quite correct, the Spack GCC's rpath *must* still be added to the executable/library, lest `libgcc_s`, `libgomp` be missing. However, this PR *does not change the behavior of GCC*: it still adds those rpath entries by default through the specs file.

The key point of this PR is that it allows the user to choose a different *mechanism* for adding the rpath to GCC's libraries:
- GCC spec file injects rpath flags (default, but prevents `LD_RUN_PATH` from working)
- Add spec suppression file so that `LD_RUN_PATH` works (requires adding `${GCC_ROOT}/lib:${GCC_ROOT}/lib64` to `LD_RUN_PATH`, easy if you add to `prefix_inspections` in Spack's `modules.yaml`)
- Add spec suppression file so that you can define `CMAKE_BUILD_RPATH` such that the GCC rpaths can be automatically removed when packaging for distribution

The third bullet is my motivating use case, but I can also imagine other users wanting the flexibility of using the second case so that `LD_RUN_PATH` actually works.

michaelkuhn commented:
Oh, sorry, I missed the explicit `--specs=norpath.spec` required for this to work.

michaelkuhn reviewed (dismissed):
LGTM but waiting for a second opinion

sethrj commented:
> Oh, sorry, I missed the explicit `--specs=norpath.spec` required for this to work.

Yeah I realize the PR description is a wall of text 😄 so it's easy to miss.

sethrj committed:
gcc: fix bad rpath on macOS

This bug has been around since the beginning of the GCC package file:
the rpath command it generates for macOS writes a single (invalid)
rpath entry.

michaelkuhn commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -643,21 +643,61 @@ def spec_dir(self):
+            # Add easily-overridable rpath string at the end
+            libdirs = [self.prefix.lib, self.prefix.lib64]
+            out.write('*link_libgcc_rpath:\n')
+            if 'platform=darwin' in spec:
```
I haven't actually run this, but I don't see `spec` being set anywhere. Shouldn't this be `self.spec`?

sethrj replied:
😱 yes... I will fix and build locally one more time to make sure no typos are committed.

sethrj committed:
gcc: only write rpaths for directories with shared libraries

The original lib64+lib was just a hack for "in case either has one" but
it's easy to tell whether either has libraries that can be directly
referenced.

sethrj commented:
By the way, the mixed toolchain of Clang CC/CXX + Gfortran only works reliably on mac because of the hard-coded `LC_ID_DYLIB` on the mac GCC libraries (linking the gfortran library, even with clang's linker, propagates its full path into the target library). I *don't* think such a toolchain will work with the "rpath in the specs file" on Linux unless you use this GCC's linker: both before and after this PR.

sethrj commented:
Thanks for the earlier catch @michaelkuhn ; I've updated the code and tested it for realsies this time.

michaelkuhn reviewed (approved):
LGTM (but waiting for someone with macOS to confirm)

sethrj commented:
@adamjstewart would you be up for confirming the macos? As I pointed out, the original version actually wrote a *garbage* rpath for mac: since `LC_RPATH` doesn't recognize colon-separated paths like Linux does.

adamjstewart commented:
How would I test this?

sethrj commented:
> How would I test this?

See if gcc builds correctly with spack? and I guess you can also check that when you link a library using `gcc` you can `grep otool -A3 RPATH foo.dylib` and make sure the path to GCC shows up there. I've done these checks myself, and as I said even if something is broken it'll be less broken than what was there before...

adamjstewart commented:
How long has GCC been broken? GCC built successfully on macOS the last time I tried (Aug 25th).

sethrj commented:
> How long has GCC been broken? GCC built successfully on macOS the last time I tried (Aug 25th).

It hasn't been  *broken* per se, it's just been adding bogus rpaths to everything it links. If you look at any executable you have linked by an existing spack-created GCC, you'll see:
```
rpath    $PREFIX/lib:$PREFIX/lib64
```
which the linker tries to evaluate as a single path rather than two separate colon-joined paths (one of which is to a nonexistent directory).

After the update, the GCC linker adds *only* the rpath to the correct lib64 directory. And since the GCC libraries have absolute path for their shared library IDs in the first path, the presence or absence of a real or fake rpath doesn't even matter.
